funsor/domains.py

```diff
@@ -2,79 +2,149 @@
+            assert cls is Real
+            assert all(isinstance(size, int) and size >= 0 for size in shape)
+            name = "Real[{}]".format(",".join(map(str, shape)))
+            result = RealType(name, (Real,), {"shape": shape})
+            Real._type_cache[shape] = result
+        return result
 
     @property
-    def size(self):
-        assert isinstance(self.dtype, int)
-        return self.dtype
+    def num_elements(cls):
+        return reduce(operator.mul, cls.shape, 1)
 
-
-@quote.register(Domain)
-def _(arg, indent, out):
-    out.append((indent, repr(arg)))
+    # DEPRECATED
+    @property
+    def dtype(self):
+        warnings.warn("domain.dtype is deprecated, "
+                      "use isinstance(domain, RealType) instead",
+                      DeprecationWarning)
+        return "real"
 
 
-def reals(*shape):
+class Real(type, metaclass=RealType):
     """
-    Construct a real domain of given shape.
+    Type of a real-valued array with known shape::
+
+        Real[()] = Real  # scalar
+        Real[8]          # vector of length 8
+        Real[3,3]        # 3x3 matrix
+
+    To dispatch on domain type, we recommend either ``@singledispatch``,
+    ``@multipledispatch``, or ``isinstance(domain, RealType)``.
     """
-    return Domain(shape, 'real')
+    _type_cache = WeakValueDictionary()
+    shape = ()
 
+    def __reduce__(self):
+        return RealType, (self.shape,)
 
-def bint(size):
+
+Real._type_cache[()] = Real  # Real[()] is Real.
+
+
+class BintType(Domain):
+    def __getitem__(cls, size):
+        # in some JAX versions, shape can be np.int64 type
+        if get_tracing_state() or funsor.get_backend() == "jax":
+            size = int(size)
+        result = Bint._type_cache.get(size, None)
+        if result is None:
+            assert cls is Bint
+            assert isinstance(size, int) and size >= 0
+            name = "Bint[{}]".format(size)
+            result = BintType(name, (Bint,), {"size": size})
+            Bint._type_cache[size] = result
+        return result
+
+    num_elements = 1
+
+    def __iter__(cls):
+        from funsor.terms import Number
+        return (Number(i, cls.dtype) for i in range(cls.size))
+
+    # DEPRECATED
+    @property
+    def dtype(cls):
+        warnings.warn("domain.dtype is deprecated, "
+                      "use isinstance(domain, BintType) instead",
+                      DeprecationWarning)
+        return cls.size
+
+    # DEPRECATED
+    @property
+    def shape(cls):
+        warnings.warn("Bint[n].shape is deprecated",
+                      DeprecationWarning)
+        return ()
+
+
+class Bint(type, metaclass=BintType):
     """
-    Construct a bounded integer domain of scalar shape.
+    Factory for bounded integer types::
+
+        Bint[5]  # integers ranging in {0,1,2,3,4}
+
+    To dispatch on domain type, we recommend either ``@singledispatch``,
+    ``@multipledispatch``, or ``isinstance(domain, BintType)``.
```

Comment:
When would we dispatch on domain type with `isinstance` rather than `issubclass`?  More broadly, when will domain types be instantiated, if ever?

Reply:
Currently we dispatch via `if domain.dtype == "real"`. I was thinking that if we are to generalize to other domain types (like `Tuple[...]`) then it would be cleanest to use `@singledispatch`. For lighter-weight uses (in my immediate refactoring plans) I will probably replace `if x.dtype == "real"` with `isinstance()`. Do you have a cleaner suggestion?

Reply:
To be clear, we would never instantiate domains like `Real`, but we do instantiate domain types like `RealType`.